cloudevents/event_v02.go

```diff
@@ -0,0 +1,260 @@
+	// Additional extension metadata beyond the base spec.
+	Extensions map[string]interface{} `json:"-,omitempty"`
+}
+
+// AsV01 implements the ContextTranslator interface.
+func (ec V02EventContext) AsV01() V01EventContext {
+	ret := V01EventContext{
+		CloudEventsVersion: V01CloudEventsVersion,
+		EventID:            ec.ID,
+		EventTime:          ec.Time,
+		EventType:          ec.Type,
+		SchemaURL:          ec.SchemaURL,
+		ContentType:        ec.ContentType,
+		Source:             ec.Source,
+		Extensions:         make(map[string]interface{}),
+	}
+	for k, v := range ec.Extensions {
+		// eventTypeVersion was retired in v0.2
+		if strings.EqualFold(k, "eventTypeVersion") {
+			etv, ok := v.(string)
+			if ok {
+				ret.EventTypeVersion = etv
+			}
+			continue
+		}
+		ret.Extensions[k] = v
+	}
+	return ret
+}
+
+// AsV02 implements the ContextTranslator interface.
+func (ec V02EventContext) AsV02() V02EventContext {
+	return ec
+}
+
+// AsHeaders implements the BinarySender interface.
+func (ec V02EventContext) AsHeaders() http.Header {
+	h := http.Header{}
+	h.Set("CE-SpecVersion", ec.SpecVersion)
+	h.Set("CE-Type", ec.Type)
+	h.Set("CE-Source", ec.Source)
+	h.Set("CE-ID", ec.ID)
+	if ec.SpecVersion == "" {
+		h.Set("CE-SpecVersion", V02CloudEventsVersion)
+	}
+	if !ec.Time.IsZero() {
+		h.Set("CE-Time", ec.Time.Format(time.RFC3339Nano))
+	}
+	if ec.SchemaURL != "" {
+		h.Set("CE-SchemaURL", ec.SchemaURL)
+	}
+	if ec.ContentType != "" {
+		h.Set("Content-Type", ec.ContentType)
+	}
+	for k, v := range ec.Extensions {
+		// Per spec, map-valued extensions are converted to a list of headers as:
+		// CE-attrib-key
+		if mapVal, ok := v.(map[string]interface{}); ok {
+			for subkey, subval := range mapVal {
+				data := fmt.Sprint(subval)
+				encoded, err := json.Marshal(subval)
+				if err != nil {
+					data = string(encoded)
+				}
+				h.Set("CE-"+k+"-"+subkey, data)
+			}
+			continue
+		}
+		data := fmt.Sprint(v)
+		encoded, err := json.Marshal(v)
+		if err != nil {
+			data = string(encoded)
+		}
+		h.Set("CE-"+k, data)
+	}
+
+	return h
+}
+
+// FromHeaders implements the BinaryLoader interface.
+func (ec *V02EventContext) FromHeaders(in http.Header) error {
+	missingField := func(name string) error {
+		if in.Get("CE-"+name) == "" {
+			return fmt.Errorf("Missing field %q in %v: %q", "CE-"+name, in, in.Get("CE-"+name))
+		}
+		return nil
+	}
+	err := anyError(
+		missingField("SpecVersion"),
+		missingField("ID"),
+		missingField("Type"),
+		missingField("Source"),
+	)
+	if err != nil {
+		return err
+	}
+	data := V02EventContext{
+		ContentType: in.Get("Content-Type"),
+		Extensions:  make(map[string]interface{}),
+	}
+	// Extensions and top-level fields are mixed under "CE-" headers.
+	// Extract them all here rather than trying to clear fields in headers.
+	for k, v := range in {
+		if strings.EqualFold(k[:len("CE-")], "CE-") {
+			key := strings.ToLower(string(k[len("CE-"):]))
+			value := v[0]
+			switch key {
```

Comment:
We can save 2 lines by:
```
switch key, value := ....;key {
...
}
```

Reply:
I'm not sure number of lines is what I'm optimizing for here. I did combine the `key, value :=` assignment.

Reply:
Well, I am mostly in favor minimal contexts/scopes for variables (which in go often coincides with brevity). But it's more a personal preference.